framework-plugins/lisk-framework-chain-connector-plugin/src/chain_connector_plugin.ts

```diff
@@ -415,104 +334,113 @@ export class ChainConnectorPlugin extends BasePlugin<ChainConnectorPluginConfig>
+			throw new Error('No certificate found in the ccu creation process.');
+		}
+		// Calculate activeValidatorsUpdate
+		let activeValidatorsUpdate: ActiveValidator[] = [];
+		let certificateThreshold = BigInt(0);
+		const blockHeader = blockHeaders.find(header => header.height === certificate.height);
 		if (!blockHeader) {
-			throw new Error('No Block header found for the given aggregate height.');
+			throw new Error('No block header found for the given certificate height.')
+		}
+		if (!this._lastCertificate.validatorsHash.equals(certificate.validatorsHash)) {
+			const validatorsUpdateResult = calculateActiveValidatorsUpdate(blockHeader, validatorsHashPreimages, this._lastCertificate);
+			activeValidatorsUpdate = validatorsUpdateResult.activeValidatorsUpdate;
+			certificateThreshold = validatorsUpdateResult.certificateThreshold;
+		}
+		// Calculate inboxUpdate
+		const inboxUpdates = await this._calculateInboxUpdate(blockHeader);
+
+		// Now create CCUs for all the inboxUpdates handling full or partial updates
+		for (const inboxUpdate of inboxUpdates) {
+			const serializedCertificate = codec.encode(certificateSchema, certificate);
+			const serializedCCUParams = codec.encode(crossChainUpdateTransactionParams, {
+				sendingChainID: this._ownChainID,
+				activeValidatorsUpdate,
+				certificate: serializedCertificate,
+				certificateThreshold,
+				inboxUpdate
+			} as CrossChainUpdateTransactionParams)
+			await this._createCCU(serializedCCUParams);
 		}
-
-		const certificate = computeCertificateFromBlockHeader(new chain.BlockHeader(blockHeader));
-		certificate.aggregationBits = blockHeader.aggregateCommit.aggregationBits;
-		certificate.signature = aggregateCommit.certificateSignature;
-
-		return certificate;
 	}
 
-	private async _checkChainOfTrust(
-		lastValidatorsHash: Buffer,
-		blsKeyToBFTWeight: Record<string, bigint>,
-		lastCertificateThreshold: bigint,
-		aggregateCommit: AggregateCommit,
-	): Promise<boolean> {
-		const blockHeaders = await this._sidechainChainConnectorDB.getBlockHeaders();
-
-		const blockHeader = blockHeaders.find(header => header.height === aggregateCommit.height - 1);
-
-		if (!blockHeader) {
-			throw new Error('No Block header found for the given aggregate height.');
+	public async unload(): Promise<void> {
+		await this._mainchainAPIClient.disconnect();
+		if (this._sidechainAPIClient) {
+			await this._sidechainAPIClient.disconnect();
 		}
 
-		// Certificate signers and certificate threshold for aggregateCommit are those authenticated by the last certificate
-		if (lastValidatorsHash === blockHeader.validatorsHash) {
-			return true;
-		}
+		this._sidechainChainConnectorDB.close();
+	}
 
-		let aggregateBFTWeight = BigInt(0);
+	private async _calculateInboxUpdate(newBlockHeader: BlockHeader): Promise<InboxUpdate[]> {
+		const crossChainMessages = await this._sidechainChainConnectorDB.getCrossChainMessages();
+		const ccmsListOfList = this._getListOfCCMs(crossChainMessages, newBlockHeader);
```

Comment:
This is a list of CCMs per CCU? Maybe can name `ccmListsPerCCU`?

Reply:
its `_groupCCMsBySize` now